lib/charms/mysql/v0/async_replication.py

```diff
@@ -0,0 +1,777 @@
+
+        return ClusterSetInstanceState(cluster_role, instance_role, relation_side)
+
+    @property
+    def cluster_name(self) -> str:
+        """This cluster name."""
+        return self._charm.app_peer_data["cluster-name"]
+
+    @property
+    def cluster_set_name(self) -> str:
+        """Cluster set name."""
+        return self._charm.app_peer_data["cluster-set-domain-name"]
+
+    def get_remote_relation_data(self, relation: Relation) -> Optional[RelationDataContent]:
+        """Remote data."""
+        if not relation.app:
+            return
+        return relation.data[relation.app]
+
+    def _on_promote_standby_cluster(self, event: ActionEvent) -> None:
+        """Promote a standby cluster to primary."""
+        if not self._charm.unit.is_leader():
+            event.fail("Only the leader unit can promote a standby cluster")
+            return
+
+        if not self._charm._mysql.is_cluster_replica():
+            event.fail("Only a standby cluster can be promoted")
+            return
+
+        if event.params.get("cluster-set-name") != self.cluster_set_name:
+            event.fail("Invalid/empty cluster set name")
+            return
+
+        # promote cluster to primary
+        cluster_name = self.cluster_name
+        force = event.params.get("force", False)
+
+        try:
+            self._charm._mysql.promote_cluster_to_primary(cluster_name, force)
+            message = f"Cluster {cluster_name} promoted to primary"
+            logger.info(message)
+            event.set_results({"message": message})
+            self._charm._on_update_status(None)
+        except MySQLPromoteClusterToPrimaryError:
+            logger.exception("Failed to promote cluster to primary")
+            event.fail("Failed to promote cluster to primary")
+
+    def _on_fence_unfence_writes_action(self, event: ActionEvent) -> None:
+        """Fence or unfence writes to a cluster."""
+        if event.params.get("cluster-set-name") != self.cluster_set_name:
+            event.fail("Invalid/empty cluster set name")
+            return
+
+        if self.role.cluster_role == "replica":
+            event.fail("Only a primary cluster can have writes fenced/unfence")
+            return
+
+        try:
+            if (
+                event.handle.kind == "fence_writes_action"
+                and not self._charm._mysql.is_cluster_writes_fenced()
+            ):
+                logger.info("Fencing writes to the cluster")
+                self._charm._mysql.fence_writes()
+                event.set_results({"message": "Writes to the cluster are now fenced"})
+            elif (
+                event.handle.kind == "unfence_writes_action"
+                and self._charm._mysql.is_cluster_writes_fenced()
+            ):
+                logger.info("Unfencing writes to the cluster")
+                self._charm._mysql.unfence_writes()
+                event.set_results({"message": "Writes to the cluster are now resumed"})
```

Comment:
should we have an else to inform the user that cluster is already fenced or cluster is already unfenced?

Reply:
nice touch eac09d4